### ISSUE TITLE
Make bonus challenges concrete

### ISSUE THREAD

**Issue by alexmojaki** [alexmojaki/futurecoder#66]
There are currently three bonus challenges in the text (search the chapters for 'bonus'). The first one is in chapter 4 and is a proper exercise with hints and a solution. The other two don't have this, they're just described in the text. They should be converted to real ExerciseSteps and moved to the end of the page with a link to the next page, just like the first one.

@spamegg1 this is a decent bite sized task, want to give it a go?

**Comment by spamegg1** [comment]
@alexmojaki Yeah sure! I was planning to tackle #34 but I'll give this a go. That one would be much longer.

**Comment by spamegg1** [comment]
The solution to the `string_sum` bonus is quite tricky: not adding the separator after the last word. They don't know list slicing, or string slicing, or `range` at this point do they?

Without those I can only come up with this somewhat hacky solution:
```python
def solution(self, words: List[str]):
    separator = ' - '
    total = ''
    total_with_separator = ''

    for word in words:
        total = total_with_separator + word
        total_with_separator = total + separator
    print(total)
```
I can explain this with hints, just making sure I'm not missing something obvious.

**Comment by alexmojaki** [comment]
What about with a boolean flag and an if statement?

**Comment by spamegg1** [comment]
OK... something like this:
```python
        def solution(self, words: List[str]):
            separator = ' - '
            total = ''
            not_finished = True

            for word in words: 
                total += word
                if not_finished:
                    total += separator

            print(total)
```

**Comment by alexmojaki** [comment]
It is tricky, which is why it's a bonus. Fortunately there's a recent exercise with some similarity.

```python
from typing import List


def solution(self, words: List[str], separator: str):
    total = ''
    not_first = False

    for word in words:
        if not_first:
            total += separator
        total += word
        not_first = True

    print(total)


solution(None, ["a", "b", "c"]," - ")
```

**Comment by spamegg1** [comment]
Ah, add the separator first... makes sense.

**Comment by spamegg1** [comment]
Ah... this is going to go horribly.
`docker-compose build` and `docker-compose up` run fine.
But the page at `localhost:3000/course` gives me:
> Method: load_data
Request data: {}
Traceback (most recent call last):
  File "/usr/src/app/backend/main/views.py", line 51, in api_view
    result = method(**args)
  File "/usr/src/app/backend/main/views.py", line 155, in load_data
    page_index=pages[self.user.page_slug].index,
KeyError: 'CombiningAndWithOr'

I think that somehow, somewhere on my PC there are some old leftover docker files with the old page names (this is happening even though I started with a fresh fork/clone).

Running `backend/manage.py test` gives
> File "/.../futurecoder/backend/book/settings.py", line 19, in <module>
    import birdseye
ModuleNotFoundError: No module named 'birdseye'

Probably some path issue.

I'll make the PR, and we'll see how it goes...

**Comment by spamegg1** [comment]
Thanks! Both issues fixed. But 1 test failed.
> First differing element 78:
{'program': ["words = ['This', 'is', 'a', '[634 chars]l)']}
{'get_solution': ['double = []', 'for numbe[445 chars]ers'}

I guess that is to be expected?

**Comment by alexmojaki** [comment]
Yes, use FIX_TESTS=1

**Comment by spamegg1** [comment]
We can close this issue.

**Comment by alexmojaki** [comment]
Yes. Next time put "Closes #66" in your PR description and it will do that automatically.
